README.md
# Asterisk cookbook

This Chef cookbook installs Asterisk either from source or packages and configures its basic settings. It also optionally installs the UniMRCP module. It is intended that this cookbook remain small and perform only installation tasks, with downstream cookbooks depending on it to configure Asterisk for more specific tasks.

# Requirements

Tested on Ubuntu 12.04, Debian 7.1 & CentOS 6.5.

# Usage

Add `recipe[asterisk]` to your node's run list. Optionally add `recipe[asterisk::unimrcp]`.

# Attributes

* `node['asterisk']['install_method']` - the method by which to install Asterisk. May be `package` or `source`. This choice determines other applicable parameters. (default `source`)
* `node['asterisk']['user']` - the user as which to run Asterisk (default `asterisk`)
* `node['asterisk']['group']` - the group as which to run Asterisk (default `asterisk`)
* `node['asterisk']['prefix']['bin']` - the prefix at which Asterisk is installed (default `/usr`)
* `node['asterisk']['prefix']['conf']` - the prefix at which Asterisk configuration is located (default `/etc`)
* `node['asterisk']['prefix']['state']` - the path at which Asterisk's runtime state located (default `/var`)

## Source install attributes
* `node['asterisk']['source']['packages']` - the packages to be installed on which compilation depends (default `%w{build-essential libssl-dev libcurl4-openssl-dev libncurses5-dev libnewt-dev libxml2-dev libsqlite3-dev uuid-dev}`)
* `node['asterisk']['source']['version']` - the version of Asterisk to install (default `11.6.0`)
* `node['asterisk']['source']['checksum']` - the checksum of the source distribution (default `80aa523bdaab71ce419022d8c851aaeda718dc5bbf5fe8cb99786da546dd8cee`)
* `node['asterisk']['source']['url']` - the url from which to download Asterisk (default `nil`)
* `node['asterisk']['source']['install_samples']` - wether or not to install sample config (default `true`)

## Package install attributes

* `node['asterisk']['package']['names']` - the Asterisk packages to install (default `%w(asterisk asterisk-dev)`)
* `node['asterisk']['package']['repo']['enable']` - if the Asterisk official repository should be enabled (default `false`)

### On Debian/Ubuntu systems:

* `node['asterisk']['package']['repo']['url']` - the URL of the Asterisk official repo (default `http://packages.asterisk.org/deb`)
* `node['asterisk']['package']['repo']['distro']` - the distro to select from the repo (default `node['lsb']['codename']`)
* `node['asterisk']['package']['repo']['branches']` - the branches of the repo to import (default `%w(main)`)
* `node['asterisk']['package']['repo']['keyserver']` - the keyserver against which to auth the repo (default `pgp.mit.edu`)
* `node['asterisk']['package']['repo']['key']` - the repo's public GPG key (default `175E41DF`)

### On RHEL/CentOS based systems:

* `node['asterisk']['package']['repo']['urls']` - the names / URLs of the Asterisk official repos (default `{
  'asterisk-11' => 'http://packages.asterisk.org/centos/$releasever/asterisk-11/$basearch/',
  'asterisk-current' => 'http://packages.asterisk.org/centos/$releasever/current/$basearch/',
}`)

## SIP attributes
* `node['asterisk']['sip']['context']` - (default `'default'`)
* `node['asterisk']['sip']['allowguest']` - (default `'yes'`)
* `node['asterisk']['sip']['allowoverlap']` - (default `'no'`)
* `node['asterisk']['sip']['allowtransfer']` - (default `'no'`)
* `node['asterisk']['sip']['realm']` - (default `'mydomain.com'`)
* `node['asterisk']['sip']['domain']` - (default `'mydomain.com'`)
* `node['asterisk']['sip']['bindport']` - (default `5060`)
* `node['asterisk']['sip']['bindaddr']` - (default `'0.0.0.0'`)
* `node['asterisk']['sip']['tcpenable']` - (default `'yes'`)
* `node['asterisk']['sip']['srvlookup']` - (default `'yes'`)
* `node['asterisk']['sip']['pedantic']` - (default `'yes'`)
* `node['asterisk']['sip']['tos_sip']` - (default `'cs3'`)
* `node['asterisk']['sip']['tos_audio']` - (default `'ef'`)
* `node['asterisk']['sip']['tos_video']` - (default `'af41'`)
* `node['asterisk']['sip']['maxexpiry']` - (default `'3600'`)
* `node['asterisk']['sip']['minexpiry']` - (default `60`)
* `node['asterisk']['sip']['defaultexpiry']` - (default `120`)
* `node['asterisk']['sip']['t1min']` - (default `100`)
* `node['asterisk']['sip']['notifymimetype']` - (default `'text/plain'`)
* `node['asterisk']['sip']['checkmwi']` - (default `10`)
* `node['asterisk']['sip']['buggymwi']` - (default `'no'`)
* `node['asterisk']['sip']['vmexten']` - (default `'voicemail'`)
* `node['asterisk']['sip']['disallow']` - (default `'all'`)
* `node['asterisk']['sip']['allow']` - (default `%w(ulaw gsm ilbc speex)`)
* `node['asterisk']['sip']['mohinterpret']` - (default `'default'`)
* `node['asterisk']['sip']['mohsuggest']` - (default `'default'`)
* `node['asterisk']['sip']['language']` - (default `'en'`)
* `node['asterisk']['sip']['relaxdtmf']` - (default `'yes'`)
* `node['asterisk']['sip']['trustrpid']` - (default `'no'`)
* `node['asterisk']['sip']['sendrpid']` - (default `'yes'`)
* `node['asterisk']['sip']['progressinband']` - (default `'never'`)
* `node['asterisk']['sip']['useragent']` - (default `'Asterisk with Adhearsion'`)
* `node['asterisk']['sip']['promiscredir']` - (default `'no'`)
* `node['asterisk']['sip']['usereqphone']` - (default `'no'`)
* `node['asterisk']['sip']['dtmfmode']` - (default `'rfc2833'`)
* `node['asterisk']['sip']['compactheaders']` - (default `'yes'`)
* `node['asterisk']['sip']['videosupport']` - (default `'yes'`)
* `node['asterisk']['sip']['maxcallbitrate']` - (default `384`)
* `node['asterisk']['sip']['callevents']` - (default `'no'`)
* `node['asterisk']['sip']['alwaysauthreject']` - (default `'yes'`)
* `node['asterisk']['sip']['g726nonstandard']` - (default `'yes'`)
* `node['asterisk']['sip']['matchexterniplocally']` - (default `'yes'`)
* `node['asterisk']['sip']['regcontext']` - (default `'sipregistrations'`)
* `node['asterisk']['sip']['rtptimeout']` - (default `60`)
* `node['asterisk']['sip']['rtpholdtimeout']` - (default `300`)
* `node['asterisk']['sip']['rtpkeepalive']` - (default `60`)
* `node['asterisk']['sip']['sipdebug']` - (default `'yes'`)
* `node['asterisk']['sip']['recordhistory']` - (default `'yes'`)
* `node['asterisk']['sip']['dumphistory']` - (default `'yes'`)
* `node['asterisk']['sip']['allowsubscribe']` - (default `'no'`)
* `node['asterisk']['sip']['subscribecontext']` - (default `'default'`)
* `node['asterisk']['sip']['notifyringing']` - (default `'yes'`)
* `node['asterisk']['sip']['notifyhold']` - (default `'yes'`)
* `node['asterisk']['sip']['limitonpeers']` - (default `'yes'`)
* `node['asterisk']['sip']['t38pt_udptl']` - (default `'yes'`)
* `node['asterisk']['public_ip']` - the public IP Asterisk listens on (default `node['ec2'] ? node['ec2']['public_ipv4'] : node['ipaddress']`)

## Manager attributes
* `node['asterisk']['manager']['enabled']` - wether or not to enable AMI (default `yes`)
* `node['asterisk']['manager']['port']` - the port on which to listen for AMI connections (default `5038`)
* `node['asterisk']['manager']['ip_address']` - the IP address on which to accept AMI connections (default `node['ec2'] ? node['ec2']['public_ipv4'] : node['ipaddress']`)
* `node['asterisk']['manager']['webenabled']` - enable AMI web connections (default `yes`)
* `node['asterisk']['manager']['timestampevents']` - wether or not to timestamp AMI events (default `yes`)
* `node['asterisk']['manager']['username']` - the username with which to authenticate AMI connections (default `manager`)
* `node['asterisk']['manager']['password']` - the password with which to authenticate AMI connections (default `password`)
* `node['asterisk']['manager']['deny']` - the ACL to deny access to (default `0.0.0.0/0.0.0.0`)
* `node['asterisk']['manager']['permit']` - the ACL to allow access to (default `127.0.0.1/255.255.255.0`)
* `node['asterisk']['manager']['read_perms']` - the AMI event classes to send to this user (default `%w(system call log verbose command agent user config)`)
* `node['asterisk']['manager']['write_perms']` - the AMI command classes to allow for this user (default `%w(system call log verbose command agent user config)`)

## UniMRCP attributes
* `node['asterisk']['unimrcp']['version']` - the version of UniMRCP to install (default `1.1.0'`)
* `node['asterisk']['unimrcp']['packages'] - the UniMRCP package dependencies to install (default %w{pkg-config build-essential}`)
* `node['asterisk']['unimrcp']['install_dir']` - the directory in which to install UniMRCP (default `/usr/local/unimrcp'`)
* `node['asterisk']['unimrcp']['server_ip']` - the IP of the MRCP server to connect to (default `127.0.0.1`)
* `node['asterisk']['unimrcp']['server_port']` - the MRCP server port to connect to (default `5060'`)
* `node['asterisk']['unimrcp']['client_ip']` - the IP of the MRCP client (default `node['ec2'] ? node['ec2']['public_ipv4'] : node['ipaddress']`)
* `node['asterisk']['unimrcp']['client_port']` - the MRCP client port (default `25097'`)
* `node['asterisk']['unimrcp']['rtp_ip']` - the client RTP IP to listen on (default `node['ec2'] ? node['ec2']['public_ipv4'] : node['ipaddress']`)
* `node['asterisk']['unimrcp']['rtp_port_min']` - the minimum RTP port (default `28000'`)
* `node['asterisk']['unimrcp']['rtp_port_max']` - the maximum RTP port (default `29000'`)

# Recipes

* `asterisk` - Fetches and installs Asterisk
* `asterisk::unimrcp` - Fetches and installs mod_unimrcp

# Author

[Ben Langfeld](@benlangfeld)



CHANGELOG.md
# develop

# 1.1.0
  * Add support for Ubuntu 14.04
  * Add support for Fedora
  * Make sure sox is installed.
  * Use correct lib_dir on RHEL/Fedora platforms, where lib != lib64
  * Use EPEL packages to install Asterisk by default
  * Fix a broken guard to allow installation from upstream packages

# 1.0.0
  * Unlock dependency on build-essential for broader compatibility
  * Update unimrcp dependency to stable release

# 0.3.0
  * CentOS support

# 0.2.7
  * Use UniMRCP cookbook to install UniMRCP and its Asterisk modules

# 0.2.6
  * Ensure sox is installed for mixing recordings

# 0.2.5
  * Allow configuring AMI event filters via attributes

# 0.2.4
  * Install latest Asterisk from source

# 0.2.3
  * Don't recompile UniMRCP after it's already successfully built

# 0.2.2
  * More sensible IP address attribute defaults

# 0.2.1
  * Fix and test building SIP peers and dialplan contexts from databags without search

# 0.2.0
  * Better testing and documentation
  * No longer test on CentOS until someone asks for it
  * Build in same directories for Package and Source installs
  * Fix builds, including UniMRCP

# 0.1.0
  * First release


recipes/source.rb
include_recipe 'apt'

include_recipe 'build-essential'

node['asterisk']['source']['packages'].each do |pkg|
  package pkg do
    options "--force-yes" if node['platform_family'] == 'debian'
  end
end

version = node['asterisk']['source']['version']
chksum = node['asterisk']['source']['checksum']
source_tarball = "asterisk-#{version}.tar.gz"
source_url_prefix = "http://downloads.asterisk.org/pub/telephony/asterisk/"
source_url_prefix << 'releases/' unless version.match(/current/)
source_url = node['asterisk']['source']['url'] ||
    source_url_prefix + source_tarball
source_path = "#{Chef::Config['file_cache_path'] || '/tmp'}/#{source_tarball}"

remote_file source_tarball do
  source source_url
  path source_path
  checksum chksum
  backup false
  notifies :create, 'ruby_block[validate asterisk tarball]', :immediately
end

# The checksum on remote_file is used only to determine if a file needs downloading
# Here we verify the checksum for security/integrity purposes
ruby_block 'validate asterisk tarball' do
  action :nothing
  block do
    require 'digest'
    expected = chksum
    actual = Digest::SHA256.file(source_path).hexdigest
    if expected and actual != expected
      raise "Checksum mismatch on #{source_path}.  Expected sha256 of #{expected} but found #{actual} instead"
    end
  end
  only_if { chksum }
end

bash "install_asterisk" do
  user "root"
  cwd File.dirname(source_path)
  code <<-EOH
    tar zxf #{source_path}
    cd asterisk-#{version =~ /(\d*)-current/ ? "#{$1}.*" : version}
    ./contrib/scripts/install_prereq install
    ./configure --prefix=#{node['asterisk']['prefix']['bin']} --sysconfdir=#{node['asterisk']['prefix']['conf']} --localstatedir=#{node['asterisk']['prefix']['state']}
    make
    make install
    make config
    #{'make samples' if node['asterisk']['source']['install_samples']}
    ldconfig
  EOH
  not_if "test -f #{node['asterisk']['prefix']['bin']}/sbin/asterisk"
  notifies :reload, resources('service[asterisk]')
end


recipes/unimrcp.rb
node.default['asterisk']['configure']['mrcp'] = true

case node['platform']
when "ubuntu","debian"
  node['asterisk']['unimrcp']['packages'].each do |pkg|
    package pkg do
      options "--force-yes"
    end
  end
end

unimrcp_name = "uni-ast-package-#{node['asterisk']['unimrcp']['version']}"
work_dir = "/tmp"
unimrcp_src_dir = "#{work_dir}/#{unimrcp_name}"

target_dir = node['asterisk']['unimrcp']['install_dir']

apr_src_dir = "#{unimrcp_src_dir}/unimrcp/libs/apr"
aprutil_src_dir = "#{unimrcp_src_dir}/unimrcp/libs/apr-util"
sofia_src_dir = "#{unimrcp_src_dir}/unimrcp/libs/sofia-sip"

remote_file "#{work_dir}/#{unimrcp_name}.tar.gz" do
  source "http://unimrcp.googlecode.com/files/#{unimrcp_name}.tar.gz"
end

bash "prepare_dir" do
  user "root"
  cwd work_dir
  code <<-EOH
    tar -zxf #{unimrcp_name}.tar.gz
  EOH
end

bash "install_apr" do
  user "root"
  cwd work_dir
  code <<-EOH
    cd #{apr_src_dir}
    ./configure --prefix=#{target_dir}
    make
    make install
  EOH
end

bash "install_apr_util" do
  user "root"
  cwd work_dir
  code <<-EOH
    cd #{aprutil_src_dir}
    ./configure --prefix=#{target_dir} --with-apr=#{apr_src_dir}
    make
    make install
  EOH
end

bash "install_sofia" do
  user "root"
  cwd work_dir
  code <<-EOH
    cd #{sofia_src_dir}
    ./configure --with-glib=no
    make
    make install
  EOH
end

bash "install_unimrcp" do
  user "root"
  cwd work_dir
  code <<-EOH
    cd #{unimrcp_src_dir}/unimrcp
    ./configure --prefix=#{target_dir} --with-apr=#{target_dir} --with-apr-util=#{target_dir}
    make
    make install
  EOH
end

directory "/var/lib/asterisk/documentation/thirdparty" do
  mode 0644
  action :create
  recursive true
end

bash "install_asterisk_modules" do
  user "root"
  cwd work_dir
  code <<-EOH
    cd #{unimrcp_src_dir}/modules
    ./configure
    make
    make install
  EOH
end

bash "ldconfig" do
  user "root"
  cwd work_dir
  code <<-EOH
    ldconfig
  EOH
end

template "/etc/asterisk/mrcp.conf" do
  source "mrcp.conf.erb"
  mode 0644
  notifies :reload, resources(:service => "asterisk")
end


attributes/default.rb
default['asterisk']['use_digium_repo']  = false
default['asterisk']['packages']         = ['asterisk', 'asterisk-dev']

#Config files we should manage
default['asterisk']['configure']['extensions'] = true
default['asterisk']['configure']['gtalk']      = true
default['asterisk']['configure']['jabber']     = true
default['asterisk']['configure']['manager']    = true
default['asterisk']['configure']['modules']    = true
default['asterisk']['configure']['mrcp']       = false
default['asterisk']['configure']['sip']        = true

# Users and dialplans.  Configrable as attributes or from data bags
default['asterisk']['users']                   = nil
default['asterisk']['auth']                    = nil
default['asterisk']['dialplan_contexts']       = nil

#Setup the Manager.conf file, refer to: http://www.voip-info.org/tiki-index.php?page=Asterisk%20config%20manager.conf
#[general]
default['asterisk']['manager_enabled']         = 'yes'
default['asterisk']['manager_port']            = 5038
default['asterisk']['manager_ip_address']      = '127.0.0.1'
default['asterisk']['manager_webenabled']      = 'yes'
default['asterisk']['manager_timestampevents'] = 'yes'

#[user] section
default['asterisk']['manager_username']    = 'manager'
default['asterisk']['manager_password']    = 'password'
default['asterisk']['manager_deny']        = '0.0.0.0/0.0.0.0'
default['asterisk']['manager_permit']      = '127.0.0.1/255.255.255.0'
default['asterisk']['manager_read_perms']  = %w(system call log verbose command agent user config)
default['asterisk']['manager_write_perms'] = %w(system call log verbose command agent user config)

#Setup the SIP.conf file, refer to: http://www.voip-info.org/wiki/view/Asterisk+config+sip.conf
default['asterisk']['sip_conf_context']              = 'default'
default['asterisk']['sip_conf_allowguest']           = 'yes'
default['asterisk']['sip_confallowoverlap']          = 'no'
default['asterisk']['sip_conf_allowtransfer']        = 'no'
default['asterisk']['sip_conf_realm']                = 'mydomain.com'
default['asterisk']['sip_conf_domain']               = 'mydomain.com'
default['asterisk']['sip_conf_bindport']             = 5060
default['asterisk']['sip_conf_bindaddr']             = '0.0.0.0'
default['asterisk']['sip_conf_tcpenable']            = 'yes'
default['asterisk']['sip_conf_srvlookup']            = 'yes'
default['asterisk']['sip_conf_pedantic']             = 'yes'
default['asterisk']['sip_conf_tos_sip']              = 'cs3'
default['asterisk']['sip_conf_tos_audio']            = 'ef'
default['asterisk']['sip_conf_tos_video']            = 'af41'
default['asterisk']['sip_conf_maxexpiry']            = '3600'
default['asterisk']['sip_conf_minexpiry']            = 60
default['asterisk']['sip_conf_defaultexpiry']        = 120
default['asterisk']['sip_conf_t1min']                = 100
default['asterisk']['sip_conf_notifymimetype']       = 'text/plain'
default['asterisk']['sip_conf_checkmwi']             = 10
default['asterisk']['sip_conf_buggymwi']             = 'no'
default['asterisk']['sip_conf_vmexten']              = 'voicemail'
default['asterisk']['sip_conf_disallow']             = 'all'
default['asterisk']['sip_conf_allow']                = %w(ulaw gsm ilbc speex)
default['asterisk']['sip_conf_mohinterpret']         = 'default'
default['asterisk']['sip_conf_mohsuggest']           = 'default'
default['asterisk']['sip_conf_language']             = 'en'
default['asterisk']['sip_conf_relaxdtmf']            = 'yes'
default['asterisk']['sip_conf_trustrpid']            = 'no'
default['asterisk']['sip_conf_sendrpid']             = 'yes'
default['asterisk']['sip_conf_progressinband']       = 'never'
default['asterisk']['sip_conf_useragent']            = 'Asterisk with Adhearsion'
default['asterisk']['sip_conf_promiscredir']         = 'no'
default['asterisk']['sip_conf_usereqphone']          = 'no'
default['asterisk']['sip_conf_dtmfmode']             = 'rfc2833'
default['asterisk']['sip_conf_compactheaders']       = 'yes'
default['asterisk']['sip_conf_videosupport']         = 'yes'
default['asterisk']['sip_conf_maxcallbitrate']       = 384
default['asterisk']['sip_conf_callevents']           = 'no'
default['asterisk']['sip_conf_alwaysauthreject']     = 'yes'
default['asterisk']['sip_conf_g726nonstandard']      = 'yes'
default['asterisk']['sip_conf_matchexterniplocally'] = 'yes'
default['asterisk']['sip_conf_regcontext']           = 'sipregistrations'
default['asterisk']['sip_conf_rtptimeout']           = 60
default['asterisk']['sip_conf_rtpholdtimeout']       = 300
default['asterisk']['sip_conf_rtpkeepalive']         = 60
default['asterisk']['sip_conf_sipdebug']             = 'yes'
default['asterisk']['sip_conf_recordhistory']        = 'yes'
default['asterisk']['sip_conf_dumphistory']          = 'yes'
default['asterisk']['sip_conf_allowsubscribe']       = 'no'
default['asterisk']['sip_conf_subscribecontext']     = 'default'
default['asterisk']['sip_conf_notifyringing']        = 'yes'
default['asterisk']['sip_conf_notifyhold']           = 'yes'
default['asterisk']['sip_conf_limitonpeers']         = 'yes'
default['asterisk']['sip_conf_t38pt_udptl']          = 'yes'

#Setup our SIP Providers
default['asterisk']['sip_providers'] = Mash.new
default['asterisk']['sip_providers']['flowroute'] = Mash.new(:type => 'friend', :host => 'sip.flowroute.com', :dtmf_mode => 'rfc2833', :context => 'flowroute', :canreinvite => 'no', :allowed_codecs => ['ulaw', 'g729'], :insecure => 'port,invite', :qualify => 'yes')

# Sensible defaults for public ip
default['asterisk']['public_ip'] = node['ec2'] ? node['ec2']['public_ipv4'] : node['ipaddress']

# UniMRCP settings
default['asterisk']['unimrcp']['version'] = '1.0.0'
default['asterisk']['unimrcp']['packages'] = %w{pkg-config build-essential}
default['asterisk']['unimrcp']['install_dir'] = '/usr/local/unimrcp'
default['asterisk']['unimrcp']['server_ip'] = '192.168.10.14'
default['asterisk']['unimrcp']['server_port'] = '5060'
default['asterisk']['unimrcp']['client_ip'] = '192.168.10.11'
default['asterisk']['unimrcp']['client_port'] = '25097'
default['asterisk']['unimrcp']['rtp_ip'] = '192.168.10.11'
default['asterisk']['unimrcp']['rtp_port_min'] = '28000'
default['asterisk']['unimrcp']['rtp_port_max'] = '29000'

#Install from source settings
default['asterisk']['source']['packages'] = %w{build-essential libssl-dev libncurses5-dev libnewt-dev libxml2-dev libsqlite3-dev uuid-dev}
default['asterisk']['source']['version'] = '11.5.1'
